crates/bytecode/src/legacy/analysis.rs

```diff
@@ -1,23 +1,37 @@
 use super::JumpTable;
 use crate::opcode;
 use bitvec::{bitvec, order::Lsb0, vec::BitVec};
-use std::sync::Arc;
+use primitives::Bytes;
+use std::{sync::Arc, vec, vec::Vec};
 
 /// Analyze the bytecode to find the jumpdests. Used to create a jump table
 /// that is needed for [`crate::LegacyAnalyzedBytecode`].
 /// This function contains a hot loop and should be optimized as much as possible.
 ///
+/// # Safety
+///
+/// The function uses unsafe pointer arithmetic, but maintains the following invariants:
+/// - The iterator never advances beyond the end of the bytecode
+/// - All pointer offsets are within bounds of the bytecode
+/// - The jump table is never accessed beyond its allocated size
+///
 /// Undefined behavior if the bytecode does not end with a valid STOP opcode. Please check
 /// [`crate::LegacyAnalyzedBytecode::new`] for details on how the bytecode is validated.
-pub fn analyze_legacy(bytetecode: &[u8]) -> JumpTable {
-    let mut jumps: BitVec<u8> = bitvec![u8, Lsb0; 0; bytetecode.len()];
+pub fn analyze_legacy(mut bytecode: Vec<u8>) -> (JumpTable, Bytes) {
+    let mut jumps: BitVec<u8> = bitvec![u8, Lsb0; 0; bytecode.len()];
 
-    let range = bytetecode.as_ptr_range();
+    if bytecode.is_empty() {
+        return (JumpTable(Arc::new(jumps)), Bytes::from(vec![opcode::STOP]));
+    }
+
+    let range = bytecode.as_ptr_range();
     let start = range.start;
     let mut iterator = start;
     let end = range.end;
+    let mut opcode = unsafe { *iterator };
```

Comment:
```suggestion
    let mut opcode = 0;
```

Nit this can be zero;